contracts/root/withdrawManager/WithdrawManager.sol

```diff
@@ -272,54 +133,107 @@ contract WithdrawManager is WithdrawManagerStorage /* , IWithdrawManager */ {
     require(ownerExits[key] == 0, "EXIT_ALREADY_IN_PROGRESS");
 
     // Calculate priority.
-    uint256 exitableAt = Math.max(_createdAt + 2 weeks, block.timestamp + 1 weeks);
+    uint256 exitableAt = Math.max(now + 2 weeks, block.timestamp + 1 weeks);
 
     PriorityQueue queue = PriorityQueue(exitsQueues[_exitObject.token]);
-    queue.insert(exitableAt, _exitId);
+    queue.insert(exitableAt, priority);
 
     // create NFT for exit UTXO
     // @todo
-    ExitNFT(exitNFTContract).mint(_exitObject.owner, _exitId);
-    exits[_exitId] = _exitObject;
+    ExitNFT(exitNFTContract).mint(_exitObject.owner, priority);
+    exits[priority] = _exitObject;
 
     // set current exit
-    ownerExits[key] = _exitId;
+    ownerExits[key] = priority;
 
     // emit exit started event
-    emit ExitStarted(_exitObject.owner, _exitId, _exitObject.token, _exitObject.receiptAmountOrNFTId);
+    emit ExitStarted(exitor, priority, rootToken, exitAmountOrTokenId, burnt);
   }
 
-  function deleteExit(uint256 exitId) external isProofValidator {
+  function addInput(uint256 exitId, uint256 age, address signer)
+    external
+  {
+    PlasmaExit storage exitObject = exits[exitId];
+    // Checks both that
+    // 1. Exit at the particular exitId exists
+    // 2. Only the predicate that started the exit is authorized to addInput 
+    require(
+      exitObject.predicate == msg.sender,
+      "EXIT_DOES_NOT_EXIST OR NOT_AUTHORIZED"
+    );
+    exitObject.inputs[age] = Input(signer);
+    emit ExitUpdated(exitId, age, signer);
+  }
+
+  function challengeExit(uint256 exitId, uint256 inputId, bytes calldata challengeData)
+    external
+  {
+    PlasmaExit storage exit = exits[exitId];
+    Input storage input = exit.inputs[inputId];
+    require(
+      exit.token != address(0x0) && input.signer != address(0x0),
+      "Invalid exit or input id"
+    );
+    bool isChallengeValid = IPredicate(exit.predicate).verifyDeprecation(
```

Comment:
``` 
require(IPredicate(exit.predicate).verifyDeprecation(
      encodeExit(exit),
      encodeInputUtxo(inputId, input),
      challengeData
    );

deleteExit(exitId);
emit ExitCancelled(exitId); 
😄

Reply:
I thought we might wanna emit an event if challenge fails. Whatchya think?

Reply:
require(condition, "msg") should be fine because no one cares about the failed challenge.